lib/obj.js

```diff
@@ -1,365 +1,479 @@
-                    result[1], result[2], undefined, result[3],
-                    result[4], result[5], undefined, result[6],
-                    result[7], result[8], undefined, result[9],
-                    result[10], result[11], undefined, result[12]
-                );
-            } else if (/^usemtl /.test(line)) {
-                var materialName = line.substring(7).trim();
-                useMaterial(materialName);
-            }
+    // Create a default node in case there are no o/g/usemtl lines in the obj
+    addNode();
+
+    // Parse the obj file
+    return readLines(objPath, parseLine)
+        .then(function() {
+            // Unload resources
+            positions = undefined;
+            normals = undefined;
+            uvs = undefined;
+
+            // Load materials and images
+            return finishLoading(nodes, mtlPaths, objPath, options);
+        });
+}
+
+function finishLoading(nodes, mtlPaths, objPath, options) {
+    nodes = cleanNodes(nodes);
+    if (nodes.length === 0) {
+        throw new RuntimeError(objPath + ' does not have any geometry data');
+    }
+    return loadMaterials(mtlPaths, objPath, options)
+        .then(function(materials) {
+            var imagePaths = getImagePaths(materials);
+            return loadImages(imagePaths, objPath, options)
+                .then(function(images) {
+                    return {
+                        nodes : nodes,
+                        materials : materials,
+                        images : images
+                    };
+                });
         });
+}
+
+function getAbsolutePath(mtlPath, objPath) {
+    if (!path.isAbsolute(mtlPath)) {
+        mtlPath = path.join(path.dirname(objPath), mtlPath);
+    }
+    return mtlPath;
+}
+
+function outsideDirectory(filePath, objPath) {
+    return (path.relative(path.dirname(objPath), filePath).indexOf('..') === 0);
+}
 
-        stream.on('end', function () {
-            resolve({
-                vertexCount: vertexCount,
-                vertexArray: vertexArray,
-                positionMin: positionMin,
-                positionMax: positionMax,
-                hasUVs: hasUVs,
-                hasNormals: hasNormals,
-                materialGroups: materialGroups,
-                materials: materials,
-                images: images
+function loadMaterials(mtlPaths, objPath, options) {
+    var materials = {};
+    return Promise.map(mtlPaths, function(mtlPath) {
+        mtlPath = getAbsolutePath(mtlPath, objPath);
+        if (options.secure && outsideDirectory(mtlPath, objPath)) {
+            options.logger('Could not read mtl file at ' + mtlPath + ' because it is outside of the obj directory and the secure flag is true. Using default material instead.');
+            return;
+        }
+        return loadMtl(mtlPath)
+            .then(function(materialsInMtl) {
+                materials = combine(materials, materialsInMtl);
+            })
+            .catch(function() {
```

Comment:
Pull `options.secure` and `options.logger` out into local variables at the start of the function.  Otherwise we re-index into options every iteration of the map.